crates/wasi/src/snapshots/preview_1.rs

```diff
@@ -0,0 +1,151 @@
+    }
+
+    fn dummy_waker() -> Waker {
+        return unsafe { Waker::from_raw(clone(5 as *const _)) };
+
+        unsafe fn clone(ptr: *const ()) -> RawWaker {
+            assert_eq!(ptr as usize, 5);
+            const VTABLE: RawWakerVTable = RawWakerVTable::new(clone, wake, wake_by_ref, drop);
+            RawWaker::new(ptr, &VTABLE)
+        }
+
+        unsafe fn wake(ptr: *const ()) {
+            assert_eq!(ptr as usize, 5);
+        }
+
+        unsafe fn wake_by_ref(ptr: *const ()) {
+            assert_eq!(ptr as usize, 5);
+        }
+
+        unsafe fn drop(ptr: *const ()) {
+            assert_eq!(ptr as usize, 5);
+        }
+    }
+}
+
+/// Creates the function item `add_wasi_snapshot_preview1_to_wasmi_linker` which when called adds all
+/// `wasi preview_1` functions to the linker
+macro_rules! impl_add_to_linker_for_funcs {
+    ($($fname:ident ($( $arg:ident : $typ:ty ),* $(,)? ) => $ret:tt),+ $(,)?) => {
+        fn add_wasi_snapshot_preview1_to_wasmi_linker<'a, T, U>(
+            linker: &mut Linker<T>,
+            mut store_ctx: impl AsContextMut<UserState = T>,
+            wasi_ctx: impl Fn(&mut T) -> &mut U + Send + Sync + Copy + 'static)
+        -> Result<(), Error>
+        where U: wasi_common::snapshots::preview_1::wasi_snapshot_preview1::WasiSnapshotPreview1 +
+                 wasi_common::snapshots::preview_1::wasi_snapshot_preview1::UserErrorConversion
+        {
+            let mut store = store_ctx.as_context_mut();
+            $(linker.define(
+                "wasi_snapshot_preview1",
+                stringify!($fname),
+                Func::wrap(&mut store, move|mut caller: Caller<'_, T>, $($arg : $typ,)*| {
+                    let result = async {
+                        let mem = match  caller.get_export("memory") {
+                            Some(Extern::Memory(m)) => m,
+                            _ => return Err(Trap::from(HostErrType::new_with_reason("missing required memory export".to_string()))),
+                        };
+                        let(mem, ctx) = mem.data_and_store_mut(&mut caller);
+                        let ctx = wasi_ctx(ctx);
+                        let mem = wiggle::wasmtime::WasmtimeGuestMemory::new(mem);
+
+                        match wasi_common::snapshots::preview_1::wasi_snapshot_preview1::$fname(ctx, &mem, $($arg,)*).await {
+                            Ok(r) => Ok(<$ret>::from(r)),
+                            Err(wiggle::Trap::String(err)) => Err(Trap::from(HostErrType::new_with_reason(err))),
+                            Err(wiggle::Trap::I32Exit(i)) => Err(Trap::from(HostErrType::new_132_exit(i))),
+
+                        }
+                    };
+                    run_in_dummy_executor(result)?
+                })
+            )?;
+
+        )+
+    Ok(())
+    }
+}
+}
+
+impl_add_to_linker_for_funcs!(
+    args_get (arg0: i32, arg1: i32) => i32,
```

Comment:
I would really like for `macro_rules` macros to have syntax that at least tries to look similar to the structures known to Rust users. For example here you are trying to define functions using their signatures. Therefore it would be great if we used Rust function declaration syntax here as much as possible.

Reply:
certainly